In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt 
import seaborn as sns

import json

import pickle

In [10]:
from fycharts.SpotifyCharts import SpotifyCharts
import sqlalchemy

In [3]:
# with open("../spotify_credentials.json", "r") as json_file:
#     creds = json.load(json_file)

In [4]:
# my_client_id = creds['SPOTIPY_CLIENT_ID']

_**NB: Ran the myCrawler.py from the terminal with the following parameters:**_

```python
from fycharts.SpotifyCharts import SpotifyCharts
import sqlalchemy

api = SpotifyCharts()
connector = sqlalchemy.create_engine("sqlite:///spotifycharts.db", echo=False)
api.viral50Daily(output_file = "viral_50_daily.csv", output_db = connector, webhook = ["https://mywebhookssite.com/post/"], start = "2019-05-01", end = "2019-05-31", region = ["it", "de"])
```

In [4]:
may_19_it_de = pd.read_csv('../viral_50_daily.csv')
may_19_it_de.head()

Position                                     Track Name         Artist  \
0         1                                        bad guy  Billie Eilish   
1         2                                       Verte Ir       DJ Luian   
2         3  Old Town Road (feat. Billy Ray Cyrus) - Remix      Lil Nas X   
3         4                                  Old Town Road      Lil Nas X   
4         5                                      Te Robaré      Nicky Jam   

         date region              spotify_id  
0  2019-05-01     it  2Fxmhks0bxGSBdJ92vM42m  
1  2019-05-01     it  4lzxJ4jCuFDXXGkE1LmpKR  
2  2019-05-01     it  6u7jPi22kF8CTQ3rb9DHE7  
3  2019-05-01     it  53CJANUxooaqGOtdsBTh7O  
4  2019-05-01     it  6v0lAdFF4haL8xjBIUjtOw

In [5]:
# api = SpotifyCharts()
# connector = sqlalchemy.create_engine("sqlite:///spotifycharts.db", echo=False)
# api.viral50Daily(output_file = "viral_50_daily.csv", output_db = connector, webhook = ["https://mywebhookssite.com/post/"], start = "2019-05-01", end = "2019-05-31", region = ["it", "de"])

It appears that I don't have to put my scraping code in a python file and run it as I did with `myCrawler.py`. I can just run the code from a jupyter notebook as I do in the cell below. 

In [15]:
# scraping Spotify API for top 200 songs in USA in May 2019
api = SpotifyCharts()
connector = sqlalchemy.create_engine("sqlite:///spotifycharts.db", echo=False)
api.top200Daily(output_file = "../data/us_may19_200.csv", output_db = connector, webhook = ["https://mywebhookssite.com/post/"], start = "2019-05-01", end = "2019-05-31", region = "us")

INFO : 09/02/2021 11:38:08 PM : Extracting top 200 daily for 2019-05-01 - us
INFO : 09/02/2021 11:38:09 PM : Extracting top 200 daily for 2019-05-02 - us
INFO : 09/02/2021 11:38:09 PM : Appending data to the table top_200_daily
INFO : 09/02/2021 11:38:09 PM : POSTing data to the endpoint https://mywebhookssite.com/post/
INFO : 09/02/2021 11:38:09 PM : Appending data to the file ../data/us_may19_200.csv...
INFO : 09/02/2021 11:38:09 PM : Done appending to the file ../data/us_may19_200.csv!!!
Exception in thread Thread-9:
Traceback (most recent call last):
  File "/Users/emilynaftalin/opt/anaconda3/lib/python3.8/site-packages/urllib3/connection.py", line 159, in _new_conn
    conn = connection.create_connection(
  File "/Users/emilynaftalin/opt/anaconda3/lib/python3.8/site-packages/urllib3/util/connection.py", line 61, in create_connection
    for res in socket.getaddrinfo(host, port, family, socket.SOCK_STREAM):
  File "/Users/emilynaftalin/opt/anaconda3/lib/python3.8/socket.py", line 9

In [16]:
# reading in csv from output file from spotify web scrape 
us_may19 = pd.read_csv('../data/us_may19_200.csv')
us_may19.head()

Position                                       Track Name         Artist  \
0         1    Old Town Road (feat. Billy Ray Cyrus) - Remix      Lil Nas X   
1         2                                          bad guy  Billie Eilish   
2         3  ME! (feat. Brendon Urie of Panic! At The Disco)   Taylor Swift   
3         4    Sunflower - Spider-Man: Into the Spider-Verse    Post Malone   
4         5                                     MIDDLE CHILD        J. Cole   

   Streams        date region              spotify_id  
0  1490233  2019-05-01     us  6u7jPi22kF8CTQ3rb9DHE7  
1  1416174  2019-05-01     us  2Fxmhks0bxGSBdJ92vM42m  
2  1407982  2019-05-01     us  4Sib57MmYGJzSvkW84jTwh  
3  1065351  2019-05-01     us  3KkXRkHbMCARz0aVfEt68P  
4   976752  2019-05-01     us  2JvzF1RMd7lE3KmFlsyZD8

In [17]:
us_may19.shape

(6200, 7)

In [49]:
# exploring how I could see how many times in a month a given song is at #1
# don't know what I would do with this information 
us_may19[ (us_may19['Track Name']=='Old Town Road (feat. Billy Ray Cyrus) - Remix')  & (us_may19['Position']==1)]

Position                                     Track Name     Artist  \
0            1  Old Town Road (feat. Billy Ray Cyrus) - Remix  Lil Nas X   
200          1  Old Town Road (feat. Billy Ray Cyrus) - Remix  Lil Nas X   
4800         1  Old Town Road (feat. Billy Ray Cyrus) - Remix  Lil Nas X   
5000         1  Old Town Road (feat. Billy Ray Cyrus) - Remix  Lil Nas X   
5200         1  Old Town Road (feat. Billy Ray Cyrus) - Remix  Lil Nas X   

      Streams        date region              spotify_id  
0     1490233  2019-05-01     us  6u7jPi22kF8CTQ3rb9DHE7  
200   1531705  2019-05-02     us  6u7jPi22kF8CTQ3rb9DHE7  
4800  1671748  2019-05-25     us  6u7jPi22kF8CTQ3rb9DHE7  
5000  1463456  2019-05-26     us  6u7jPi22kF8CTQ3rb9DHE7  
5200  1439533  2019-05-27     us  6u7jPi22kF8CTQ3rb9DHE7

In [47]:
# everything I need to interact with Spotify API and thereby get the audio features

import spotipy
from spotipy.oauth2 import SpotifyClientCredentials

with open("../spotify_credentials.json", "r") as json_file:
    creds = json.load(json_file)

my_client_id = creds['SPOTIPY_CLIENT_ID']
my_client_secret = creds['SPOTIPY_CLIENT_SECRET']

client_credentials_manager = SpotifyClientCredentials(client_id=my_client_id, client_secret=my_client_secret)
sp = spotipy.Spotify(client_credentials_manager=client_credentials_manager)

In [48]:
# getting all the audio features for the us_May19 tracks 

rows = []          #list to append each track features to 
batchsize = 100    #max number of track ids we're allowed to submit per query
None_counter = 0   #count if there are any songs without any audio features

for i in range(0,len(us_may19['spotify_id']), batchsize):    #batchsize = offset
    batch = us_may19['spotify_id'][i:i+batchsize]            #batch is now a list of all 929 track_ids
    
    feature_results = sp.audio_features(batch)              #begins querying the audio features endpoint

    for i, t in enumerate(feature_results):
        if t == None:                               #if the audio features for a song are missing, count 1        
            None_counter += 1          
        else:
            rows.append(t)  
print('Number of tracks where no audio features were available:',None_counter)
print('Number of usable tracks:', len(rows))

Number of tracks where no audio features were available: 0
Number of usable tracks: 6200


In [51]:
type(rows)

list

In [53]:
# building dataframe from rows, which is a list of dictionaries (each dictionary represents the audio features from one song)
feat_us_may19 = pd.DataFrame(rows)
feat_us_may19.head()

danceability  energy  key  loudness  mode  speechiness  acousticness  \
0         0.878   0.619    6    -5.560     1       0.1020        0.0533   
1         0.701   0.425    7   -10.965     1       0.3750        0.3280   
2         0.614   0.839    0    -4.143     1       0.0702        0.0291   
3         0.760   0.479    2    -5.574     1       0.0466        0.5560   
4         0.837   0.364    8   -11.713     1       0.2760        0.1490   

   instrumentalness  liveness  valence    tempo            type  \
0              0.00    0.1130    0.639  136.041  audio_features   
1              0.13    0.1000    0.562  135.128  audio_features   
2              0.00    0.1120    0.657  182.123  audio_features   
3              0.00    0.0703    0.913   89.911  audio_features   
4              0.00    0.2710    0.463  123.984  audio_features   

                       id                                   uri  \
0  6u7jPi22kF8CTQ3rb9DHE7  spotify:track:6u7jPi22kF8CTQ3rb9DHE7   
1  2Fxmhks0bxGSBdJ92vM42m  spotify:track:2Fxmhks0bxGSBdJ92vM42m   
2  4Sib57MmYGJzSvkW84jTwh  spotify:track:4Sib57MmYGJzSvkW84jTwh   
3  3KkXRkHbMCARz0aVfEt68P  spotify:track:3KkXRkHbMCARz0aVfEt68P   
4  2JvzF1RMd7lE3KmFlsyZD8  spotify:track:2JvzF1RMd7lE3KmFlsyZD8   

                                          track_href  \
0  https://api.spotify.com/v1/tracks/6u7jPi22kF8C...   
1  https://api.spotify.com/v1/tracks/2Fxmhks0bxGS...   
2  https://api.spotify.com/v1/tracks/4Sib57MmYGJz...   
3  https://api.spotify.com/v1/tracks/3KkXRkHbMCAR...   
4  https://api.spotify.com/v1/tracks/2JvzF1RMd7lE...   

                                        analysis_url  duration_ms  \
0  https://api.spotify.com/v1/audio-analysis/6u7j...       157067   
1  https://api.spotify.com/v1/audio-analysis/2Fxm...       194088   
2  https://api.spotify.com/v1/audio-analysis/4Sib...       193040   
3  https://api.spotify.com/v1/audio-analysis/3KkX...       158040   
4  https://api.spotify.com/v1/audio-analysis/2Jvz...       213594   

   time_signature  
0               4  
1               4  
2               4  
3               4  
4               4

In [54]:
# merging chart songs dataframe with audio features dataframe 

us_may_19_df = pd.concat([us_may19, feat_us_may19], axis=1)
us_may_19_df.head()

Position                                       Track Name         Artist  \
0         1    Old Town Road (feat. Billy Ray Cyrus) - Remix      Lil Nas X   
1         2                                          bad guy  Billie Eilish   
2         3  ME! (feat. Brendon Urie of Panic! At The Disco)   Taylor Swift   
3         4    Sunflower - Spider-Man: Into the Spider-Verse    Post Malone   
4         5                                     MIDDLE CHILD        J. Cole   

   Streams        date region              spotify_id  danceability  energy  \
0  1490233  2019-05-01     us  6u7jPi22kF8CTQ3rb9DHE7         0.878   0.619   
1  1416174  2019-05-01     us  2Fxmhks0bxGSBdJ92vM42m         0.701   0.425   
2  1407982  2019-05-01     us  4Sib57MmYGJzSvkW84jTwh         0.614   0.839   
3  1065351  2019-05-01     us  3KkXRkHbMCARz0aVfEt68P         0.760   0.479   
4   976752  2019-05-01     us  2JvzF1RMd7lE3KmFlsyZD8         0.837   0.364   

   key  ...  liveness  valence    tempo            type  \
0    6  ...    0.1130    0.639  136.041  audio_features   
1    7  ...    0.1000    0.562  135.128  audio_features   
2    0  ...    0.1120    0.657  182.123  audio_features   
3    2  ...    0.0703    0.913   89.911  audio_features   
4    8  ...    0.2710    0.463  123.984  audio_features   

                       id                                   uri  \
0  6u7jPi22kF8CTQ3rb9DHE7  spotify:track:6u7jPi22kF8CTQ3rb9DHE7   
1  2Fxmhks0bxGSBdJ92vM42m  spotify:track:2Fxmhks0bxGSBdJ92vM42m   
2  4Sib57MmYGJzSvkW84jTwh  spotify:track:4Sib57MmYGJzSvkW84jTwh   
3  3KkXRkHbMCARz0aVfEt68P  spotify:track:3KkXRkHbMCARz0aVfEt68P   
4  2JvzF1RMd7lE3KmFlsyZD8  spotify:track:2JvzF1RMd7lE3KmFlsyZD8   

                                          track_href  \
0  https://api.spotify.com/v1/tracks/6u7jPi22kF8C...   
1  https://api.spotify.com/v1/tracks/2Fxmhks0bxGS...   
2  https://api.spotify.com/v1/tracks/4Sib57MmYGJz...   
3  https://api.spotify.com/v1/tracks/3KkXRkHbMCAR...   
4  https://api.spotify.com/v1/tracks/2JvzF1RMd7lE...   

                                        analysis_url duration_ms  \
0  https://api.spotify.com/v1/audio-analysis/6u7j...      157067   
1  https://api.spotify.com/v1/audio-analysis/2Fxm...      194088   
2  https://api.spotify.com/v1/audio-analysis/4Sib...      193040   
3  https://api.spotify.com/v1/audio-analysis/3KkX...      158040   
4  https://api.spotify.com/v1/audio-analysis/2Jvz...      213594   

  time_signature  
0              4  
1              4  
2              4  
3              4  
4              4  

[5 rows x 25 columns]

In [56]:
us_may_19_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6200 entries, 0 to 6199
Data columns (total 25 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Position          6200 non-null   int64  
 1   Track Name        6200 non-null   object 
 2   Artist            6200 non-null   object 
 3   Streams           6200 non-null   int64  
 4   date              6200 non-null   object 
 5   region            6200 non-null   object 
 6   spotify_id        6200 non-null   object 
 7   danceability      6200 non-null   float64
 8   energy            6200 non-null   float64
 9   key               6200 non-null   int64  
 10  loudness          6200 non-null   float64
 11  mode              6200 non-null   int64  
 12  speechiness       6200 non-null   float64
 13  acousticness      6200 non-null   float64
 14  instrumentalness  6200 non-null   float64
 15  liveness          6200 non-null   float64
 16  valence           6200 non-null   float64


In [55]:
us_may_19_df['time_signature'].value_counts()

4    5995
3     109
5      83
1      13
Name: time_signature, dtype: int64

In [57]:
us_may_19_df['danceability'].mean()

0.7103354838709686

In [74]:
# artists danceability scores (mean across all their songs)
us_may_19_df.groupby('Artist')[['danceability']].mean().sort_values(by='danceability', ascending=False)[:20]

danceability
Artist                              
Tory Lanez                  0.962406
Rich The Kid                0.942000
City Girls                  0.938000
Ski Mask The Slump God      0.935000
6ix9ine                     0.931000
BlocBoy JB                  0.922000
Megan Thee Stallion         0.921000
Russ                        0.913000
Kendrick Lamar              0.908000
Chance the Rapper           0.906000
Blueface                    0.906000
Metro Boomin                0.901000
Offset                      0.900087
ScHoolboy Q                 0.898167
Lil Baby                    0.894372
Lil Nas X                   0.893000
Meek Mill                   0.889000
NLE Choppa                  0.888000
Tyga                        0.884000
YG                          0.879103

In [65]:
len(us_may_19_df)

6200

In [66]:
len(us_may_19_df.groupby('Artist')[['danceability']].mean().sort_values(by='danceability', ascending=False))

164

In [73]:
# times in the top 200 in May 
us_may_19_df.groupby('Artist').count().sort_values(by='Position', ascending=False)[:15]

Position  Track Name  Streams  date  region  spotify_id  \
Artist                                                                        
Billie Eilish            459         459      459   459     459         459   
Juice WRLD               261         261      261   261     261         261   
Post Malone              254         254      254   254     254         254   
Khalid                   217         217      217   217     217         217   
Tyler, The Creator       158         158      158   158     158         158   
XXXTENTACION             156         156      156   156     156         156   
Ariana Grande            129         129      129   129     129         129   
Logic                    124         124      124   124     124         124   
Lil Baby                 113         113      113   113     113         113   
Drake                    111         111      111   111     111         111   
Offset                   104         104      104   104     104         104   
DJ Khaled                 95          95       95    95      95          95   
Cardi B                   94          94       94    94      94          94   
Lil Uzi Vert              93          93       93    93      93          93   
Luke Combs                88          88       88    88      88          88   

                    danceability  energy  key  loudness  ...  liveness  \
Artist                                                   ...             
Billie Eilish                459     459  459       459  ...       459   
Juice WRLD                   261     261  261       261  ...       261   
Post Malone                  254     254  254       254  ...       254   
Khalid                       217     217  217       217  ...       217   
Tyler, The Creator           158     158  158       158  ...       158   
XXXTENTACION                 156     156  156       156  ...       156   
Ariana Grande                129     129  129       129  ...       129   
Logic                        124     124  124       124  ...       124   
Lil Baby                     113     113  113       113  ...       113   
Drake                        111     111  111       111  ...       111   
Offset                       104     104  104       104  ...       104   
DJ Khaled                     95      95   95        95  ...        95   
Cardi B                       94      94   94        94  ...        94   
Lil Uzi Vert                  93      93   93        93  ...        93   
Luke Combs                    88      88   88        88  ...        88   

                    valence  tempo  type   id  uri  track_href  analysis_url  \
Artist                                                                         
Billie Eilish           459    459   459  459  459         459           459   
Juice WRLD              261    261   261  261  261         261           261   
Post Malone             254    254   254  254  254         254           254   
Khalid                  217    217   217  217  217         217           217   
Tyler, The Creator      158    158   158  158  158         158           158   
XXXTENTACION            156    156   156  156  156         156           156   
Ariana Grande           129    129   129  129  129         129           129   
Logic                   124    124   124  124  124         124           124   
Lil Baby                113    113   113  113  113         113           113   
Drake                   111    111   111  111  111         111           111   
Offset                  104    104   104  104  104         104           104   
DJ Khaled                95     95    95   95   95          95            95   
Cardi B                  94     94    94   94   94          94            94   
Lil Uzi Vert             93     93    93   93   93          93            93   
Luke Combs               88     88    88   88   88          88            88   

                    duration_ms  time_signature  
Artist

In [71]:
us_may_19_df['Artist'].value_counts()

Billie Eilish         459
Juice WRLD            261
Post Malone           254
Khalid                217
Tyler, The Creator    158
                     ... 
Jon Bellion             1
Bastille                1
6ix9ine                 1
A$AP Ferg               1
Chelsea Cutler          1
Name: Artist, Length: 164, dtype: int64

_So there were 164 artists represented on the US top 200 chart in May 2019._

In [75]:
us_may_19_df['type'].value_counts()

audio_features    6200
Name: type, dtype: int64